stl/src/atomic_wait.cpp

```diff
@@ -0,0 +1,340 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// implement atomic wait / notify_one / notify_all
+
+// clang-format off
+
+#include <atomic>
+#include <cstdint>
+#include <new>
+#include <thread>
+#include <Windows.h>
+#include <VersionHelpers.h>
+
+// clang-format on
+
+namespace {
+
+    constexpr size_t _Wait_table_size_power = 8;
+    constexpr size_t _Wait_table_size       = 1 << _Wait_table_size_power;
+    constexpr size_t _Wait_table_index_mask = _Wait_table_size - 1;
+
+#pragma warning(push)
+#pragma warning(disable : 4324) // structure was padded due to alignment specifier
+    struct alignas(std::hardware_destructive_interference_size) _Wait_table_entry {
+        // Arbitrary variable to wait/notify on if target variable is not proper atomic for that
+        // Size is largest of lock-free to make aliasing problem into hypothetical
+        std::atomic<unsigned long long> _Counter{};
+
+        CONDITION_VARIABLE _Condition = CONDITION_VARIABLE_INIT;
+        SRWLOCK _Lock                 = SRWLOCK_INIT;
+
+        constexpr _Wait_table_entry() noexcept = default;
+    };
+#pragma warning(pop)
+
+    _Wait_table_entry& _Atomic_wait_table_entry(const void* const _Storage) noexcept {
+        static _Wait_table_entry wait_table[_Wait_table_size];
+        auto index = reinterpret_cast<std::uintptr_t>(_Storage);
+        index ^= index >> (_Wait_table_size_power * 2);
+        index ^= index >> _Wait_table_size_power;
+        return wait_table[index & _Wait_table_index_mask];
+    }
+
+    unsigned long _Get_remaining_waiting_time(_Atomic_wait_context_t& _Wait_context) {
+        const unsigned long long deadline = _Wait_context._Deadline;
+        if (deadline == _Atomic_wait_no_deadline) {
+            return INFINITE;
+        }
+
+        const unsigned long long current_time = ::GetTickCount64();
+        if (current_time >= deadline) {
+            return 0;
+        }
+
+        unsigned long long remaining      = deadline - current_time;
+        constexpr unsigned long _Ten_days = 864'000'000;
+        if (remaining > _Ten_days) {
+            return _Ten_days;
+        }
+        return static_cast<unsigned long>(remaining);
+    }
+
+    void _Assume_timeout() noexcept {
+#ifndef _NDEBUG

```

Comment:
`_NDEBUG` isn't a macro. `NDEBUG` controls the behavior of the `assert` macro (and the STL otherwise never interacts with this). `_DEBUG` is implied by `/MDd` (dynamic debug) and `/MTd` (static debug) and indicates whether we're in "debug mode".

Reply:
Changed to `#ifdef _NDEBUG`. Let me know if you see enough value in this check to enable it in release.